### PR TITLE
Add a flag that checkerboards offscreen render targets created by SkCanvas.saveLayer

### PR DESCRIPTION
**Pull request by jason-simmons**
Fixes https://github.com/flutter/flutter/issues/9473

### CODE REVIEW

**Comment by jason-simmons** [comment]
@chinmaygarde

File: flow/layers/clip_rrect_layer.cc [hunk 1]
```diff
@@ -37,8 +37,7 @@ void ClipRRectLayer::Paint(PaintContext& context) {
   TRACE_EVENT0("flutter", "ClipRRectLayer::Paint");
   FTL_DCHECK(!needs_system_composite());
 
-  SkAutoCanvasRestore save(&context.canvas, false);
-  context.canvas.saveLayer(&paint_bounds(), nullptr);
+  Layer::AutoSaveLayer save(context, paint_bounds(), nullptr);
```
Comment (chinmaygarde) [inline flow/layers/clip_rrect_layer.cc]: In the previous version of this code, we were doing a `SkCanvas::save (via SkAutoCanvasRestore)` -> `SkCanvas::saveLayer` -> PaintChildren -> `SkCanvas::restore (via SkAutoCanvasRestore)`. In this version, we no longer have the `SkCanvas::save`. Is this fine? I am not sure why we did the SkCanvas::save in the first place.

File: flow/layers/layer.h [hunk 2]
```diff
@@ -45,6 +45,25 @@ class Layer {
...
+                  const SkPaint* paint);
+
+    AutoSaveLayer(const PaintContext& paint_context,
+                  const SkCanvas::SaveLayerRec& layer_rec);
+
+    ~AutoSaveLayer();
+
+   private:
+    const PaintContext& paint_context_;
+    const SkRect bounds_;
```
Comment (chinmaygarde) [inline flow/layers/layer.h]: `FTL_DISALLOW_COPY_AND_ASSIGN`.

Comment (chinmaygarde) [inline flow/layers/layer.h]: Wait, nevermind. Didn't notice the `const`s.

File: flow/layers/layer_tree.cc [hunk 3]
```diff
@@ -52,7 +53,8 @@ void LayerTree::UpdateScene(SceneUpdateContext& context,
 void LayerTree::Paint(CompositorContext::ScopedFrame& frame) {
   Layer::PaintContext context = {frame.canvas(), frame.context().frame_time(),
                                  frame.context().engine_time(),
-                                 frame.context().memory_usage()};
+                                 frame.context().memory_usage(),
+                                 checkerboard_offscreen_layers_};
```
Comment (chinmaygarde) [inline flow/layers/layer_tree.cc]: I don't notice similar updates to `scene_update_context.cc`. Are you sure you checked the Fuchsia build?

**Comment by chinmaygarde** [comment]
Travis is mad about paint_utils in the licenses.

File: flow/layers/clip_rrect_layer.cc (hunk 1 above)
Comment (jason-simmons) [inline flow/layers/clip_rrect_layer.cc]: SkAutoCanvasRestore(canvas, false) doesn't do an SkCanvas::save().  The SkCanvas::save() happens within the saveLayer() call

File: flow/layers/layer_tree.cc (hunk 3 above)
Comment (jason-simmons) [inline flow/layers/layer_tree.cc]: done

**Pull request merged**